File: labs/lab2_ws/src/safety_node/src/safety_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
#include <vector>
#include <algorithm>
#include <cmath>


class Safety : public rclcpp::Node {
    // The class that handles emergency braking

public:
    Safety() : Node("Safety_node"), speed_(0.0), closest_obstacle_distance_(std::numeric_limits<double>::infinity()), collision_threshold_(0.5) 
    {
        /*
        You should also subscribe to the /scan topic to get the
        sensor_msgs/LaserScan messages and the /ego_racecar/odom topic to get
        the nav_msgs/Odometry messages

        The subscribers should use the provided odom_callback and
        scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        */
       drive_publisher_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("/drive", 10);
       scan_subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>("/scan", 10, std::bind(&Safety::scan_callback, this, std::placeholders::_1));
       odom_subscription_ = this->create_subscription<nav_msgs::msg::Odometry>("/ego_racecar/odom", 10, std::bind(&Safety::odom_callback, this, std::placeholders::_1));

       RCLCPP_INFO(this->get_logger(), "Safety node has been started");
    }

private:
    void odom_callback(const nav_msgs::msg::Odometry::ConstSharedPtr msg)
    {
        speed_ = msg->twist.twist.linear.x;
        RCLCPP_INFO(this->get_logger(), "Current Speed: %.2f m/s", speed_);
    }

    void scan_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr msg)
    {
        std::vector<float> ranges = msg->ranges;
        std::replace_if(ranges.begin(), ranges.end(), [](float r) { return !std::isfinite(r); }, std::numeric_limits<float>::infinity());

        auto min_element_it = std::min_element(ranges.begin(), ranges.end());
        closest_obstacle_distance_ = (min_element_it != ranges.end()) ? *min_element_it : std::numeric_limits<double>::infinity();
        RCLCPP_INFO(this->get_logger(), "Closest Obstacle Distance: %.2f meters", closest_obstacle_distance_);

        std::vector<double> angles;
        for (size_t i = 0; i < ranges.size(); ++i) {
            angles.push_back(msg->angle_min + i * msg->angle_increment);
        }

        std::vector<double> range_rates;
        for (double angle : angles) {
            range_rates.push_back(speed_ * std::cos(angle));
        }

        std::vector<double> ittc_values;
        for (size_t i = 0; i < ranges.size(); ++i) {
            if (range_rates[i] < 0) {
                ittc_values.push_back(ranges[i] / std::max(-range_rates[i], 1e-6));
            } else {
                ittc_values.push_back(std::numeric_limits<double>::infinity());
            }
        }

        if (std::any_of(ittc_values.begin(), ittc_values.end(), [this](double ttc) { return ttc < collision_threshold_; })) {
            RCLCPP_WARN(this->get_logger(), "COLLISION WARNING: iTTC below threshold! Braking...");
            publish_drive_command(0.0, 0.0);
        }
    }

    void publish_drive_command(double speed, double steering_angle) {
        auto drive_msg = ackermann_msgs::msg::AckermannDriveStamped();
        drive_msg.drive.speed = speed;
        drive_msg.drive.steering_angle = steering_angle;
        drive_publisher_->publish(drive_msg);

        RCLCPP_INFO(this->get_logger(), "Published: Speed=%.2f, Steering Angle=%.2f", speed, steering_angle);
    }

    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_publisher_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_subscription_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_subscription_;

    double speed_;
    double closest_obstacle_distance_;
    double collision_threshold_;
};

int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<Safety>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: labs/lab2_ws/src/safety_node/scripts/safety_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

import numpy as np
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import Odometry
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive


class SafetyNode(Node):
    """
    The class that handles emergency braking.
    """
    def __init__(self):
        super().__init__('safety_node')
        """
        One publisher should publish to the /drive topic with a AckermannDriveStamped drive message.

        You should also subscribe to the /scan topic to get the LaserScan messages and
        the /ego_racecar/odom topic to get the current speed of the vehicle.

        The subscribers should use the provided odom_callback and scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        """
        self.drive_publisher = self.create_publisher(
            AckermannDriveStamped,
            '/drive',
            10
        )

        self.odom_subscription = self.create_subscription(
            Odometry,
            '/ego_racecar/odom',
            self.odom_callback,
            10
        )

        self.scan_subscription = self.create_subscription(
            LaserScan,
            '/scan',
            self.scan_callback,
            10
        )

        # Vehicle state variables
        self.speed = 0.
        self.closet_obstacle_distance = float('inf')
        self.prev_ranges = None # Store previous range measurements
        self.prev_scan_time = None # Store previous scan timestamp
        self.collision_threshold = 0.5 # Threshold for iTTC warning

        self.get_logger().info('Safety Node Initialized')


    def odom_callback(self, odom_msg):
        """
        Process Odometry messages to get current vehicle speed.
        """
        self.speed = odom_msg.twist.twist.linear.x

        self.get_logger().info(f'Current Speed: {self.speed:.2f} m/s')


    def scan_callback(self, scan_msg):
        """
        Process LaserScan message to determine closest obstacle distance.
        """
        ranges = np.array(scan_msg.ranges)
        ranges = np.where(np.isfinite(ranges), ranges, float('inf')) # Handle NaNs and Infs

        self.closest_obstacle_distance = np.min(ranges) # Find the minimum range value

        self.get_logger().info(f'Furthest Obstacle Distance: {np.max(ranges):.2f} meters')
        self.get_logger().info(f'Closest Obstacle Distance: {self.closest_obstacle_distance:.2f} meters')

        # angles = np.linspace(scan_msg.angle_min, scan_msg.angle_max, len(ranges))
        angles = np.arange(scan_msg.angle_min, scan_msg.angle_max, scan_msg.angle_increment)

        # Compute range rate using velocity projection
        range_rates = self.speed * np.cos(angles)
        scan_time = scan_msg.header.stamp.sec + scan_msg.header.stamp.nanosec * 1e-9

        # Compute range rate using finite difference if previous scan is available
        if self.prev_ranges is not None and self.prev_scan_time is not None:
            delta_t = scan_time - self.prev_scan_time
            
            if delta_t > 0:
                finite_diff_rates = -(ranges - self.prev_ranges) / delta_t
                range_rates = np.where(finite_diff_rates < 0, finite_diff_rates, range_rates) # Use meaningful rate

        # Compute iTTC: iTTC = r / max(-r_dot, 0)
        range_rates[range_rates >= 0] = float('-inf') # Ensure valid calculations
        ittc_values = ranges / np.maximum(-range_rates, 1e-6) # Avoid division by zero

        self.get_logger().info(f'Range Rate Max Value: {np.max(range_rates)} / Range Rate Min Value: {np.min(range_rates)}')  
        self.get_logger().info(f'iTTC Max Value: {np.max(ittc_values)} / iTTC Min Value: {np.min(ittc_values)}')       

        # Check for imminent collisions
        if np.any(ittc_values < self.collision_threshold):
            self.get_logger().warn("COLLISION WARNIN: iTTC below threshold! Braking...")
            # self.publish_drive_command(0.0, 0.0) # Hald vehicle
        
        # Store current scan for next iteration
        self.prev_ranges = ranges
        self.prev_scan_time = scan_time


    def publish_drive_command(self, speed, steering_angle):
        """
        Publishes AckermannDriveStamped message to control the vehicle.
        """
        drive_msg = AckermannDriveStamped()
        drive_msg.drive.speed = speed
        drive_msg.drive.steering_angle = steering_angle
        
        self.drive_publisher.publish(drive_msg)

        self.get_logger().info(f'Published: Speed={speed:.2f}, Steering Angle={steering_angle:.2f}')


# Destory the node explicitly
# (optional - otherwise it will be done automatically
# when the garbage collector destroys the node object)
def main(args=None):
    rclpy.init(args=args)
    safety_node = SafetyNode()

    try:
        rclpy.spin(safety_node)
    except KeyboardInterrupt:
        pass
    finally:
        safety_node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
